// appium-ios-driver/build/lib/commands/logging.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _appiumIosLog = require('appium-ios-log');

var _logger = require('../logger');

var _logger2 = _interopRequireDefault(_logger);

var commands = {},
    helpers = {},
    extensions = {};

var SUPPORTED_LOG_TYPES = {
  'syslog': 'System Logs - Device logs for iOS applications on real devices and simulators',
  'crashlog': 'Crash Logs - Crash reports for iOS applications on real devices and simulators',
  'performance': 'Performance Logs - Debug Timelines on real devices and simulators'
};

commands.getLogTypes = function callee$0$0() {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        _logger2['default'].debug('Retrieving supported log types');
        return context$1$0.abrupt('return', _lodash2['default'].keys(SUPPORTED_LOG_TYPES));

      case 2:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

commands.getLog = function callee$0$0(logType) {
  var logs;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        _logger2['default'].debug('Retrieving \'' + logType + '\' logs');
        // Check if passed logType is supported

        if (_lodash2['default'].has(SUPPORTED_LOG_TYPES, logType)) {
          context$1$0.next = 3;
          break;
        }

        throw new Error('Unsupported log type \'' + logType + '\' for this device. Supported types : ' + JSON.stringify(SUPPORTED_LOG_TYPES));

      case 3:
        if (!_lodash2['default'].isEmpty(this.logs)) {
          context$1$0.next = 5;
          break;
        }

        throw new Error('No logs currently available. Is the device/simulator started?');

      case 5:
        context$1$0.next = 7;
        return _regeneratorRuntime.awrap(this.logs[logType].getLogs());

      case 7:
        logs = context$1$0.sent;

        if (!logs) {
          context$1$0.next = 12;
          break;
        }

        return context$1$0.abrupt('return', logs);

      case 12:
        throw new Error('No logs of type \'' + logType + '\' found.');

      case 13:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

helpers.startLogCapture = function callee$0$0(sim) {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (_lodash2['default'].isEmpty(this.logs)) {
          context$1$0.next = 3;
          break;
        }

        _logger2['default'].warn("Trying to start iOS log capture but it's already started!");
        return context$1$0.abrupt('return');

      case 3:
        this.logs.crashlog = new _appiumIosLog.IOSCrashLog();
        this.logs.syslog = new _appiumIosLog.IOSLog({
          sim: sim,
          udid: this.opts.udid,
          showLogs: this.opts.showIOSLog
        });
        context$1$0.prev = 5;
        context$1$0.next = 8;
        return _regeneratorRuntime.awrap(this.logs.syslog.startCapture());

      case 8:
        context$1$0.next = 14;
        break;

      case 10:
        context$1$0.prev = 10;
        context$1$0.t0 = context$1$0['catch'](5);

        _logger2['default'].warn("Could not capture logs from device. Continuing without capturing logs.");
        return context$1$0.abrupt('return');

      case 14:
        context$1$0.next = 16;
        return _regeneratorRuntime.awrap(this.logs.crashlog.startCapture());

      case 16:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[5, 10]]);
};

_Object$assign(extensions, commands, helpers);
exports.commands = commands;
exports.helpers = helpers;
exports.SUPPORTED_LOG_TYPES = SUPPORTED_LOG_TYPES;
exports['default'] = extensions;

// make sure that we have logs at all
// otherwise it's not been initialized

// If logs captured successfully send response with data, else send error
