// appium-ios-driver/build/index.js
#!/usr/bin/env node

require('source-map-support').install();

'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

var _interopRequireWildcard = require('babel-runtime/helpers/interop-require-wildcard')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _yargs = require('yargs');

var _yargs2 = _interopRequireDefault(_yargs);

var _asyncbox = require('asyncbox');

var _libServer = require('./lib/server');

var _libDriver = require('./lib/driver');

var _libDesiredCaps = require('./lib/desired-caps');

var _libCommandsIndex = require('./lib/commands/index');

var _libSettings = require('./lib/settings');

var settings = _interopRequireWildcard(_libSettings);

var _libDevice = require('./lib/device');

var device = _interopRequireWildcard(_libDevice);

var _libUtils = require('./lib/utils');

var _libUtils2 = _interopRequireDefault(_libUtils);

var DEFAULT_HOST = "localhost";
var DEFAULT_PORT = 4723;

function main() {
  var port, host;
  return _regeneratorRuntime.async(function main$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        port = _yargs2['default'].argv.port || DEFAULT_PORT;
        host = _yargs2['default'].argv.host || DEFAULT_HOST;
        return context$1$0.abrupt('return', (0, _libServer.startServer)(port, host));

      case 3:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

if (require.main === module) {
  (0, _asyncbox.asyncify)(main);
}

exports.IosDriver = _libDriver.IosDriver;
exports.desiredCapConstraints = _libDesiredCaps.desiredCapConstraints;
exports.desiredCapValidation = _libDesiredCaps.desiredCapValidation;
exports.commands = _libCommandsIndex.commands;
exports.iosCommands = _libCommandsIndex.iosCommands;
exports.settings = settings;
exports.device = device;
exports.defaultServerCaps = _libDriver.defaultServerCaps;
exports.utils = _libUtils2['default'];
exports['default'] = _libDriver.IosDriver;
