// appium-xcuitest-driver/build/lib/commands/source.js
'use strict';

var _defineProperty = require('babel-runtime/helpers/define-property')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _js2xmlparser2 = require("js2xmlparser2");

var _js2xmlparser22 = _interopRequireDefault(_js2xmlparser2);

var commands = {},
    helpers = {},
    extensions = {};

// WDA uses a POST request to get the source. Until that
// is fixed, we need to translate the request
commands.getPageSource = function callee$0$0() {
  var cmd;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!this.isWebContext()) {
          context$1$0.next = 5;
          break;
        }

        cmd = 'document.getElementsByTagName("html")[0].outerHTML';
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap(this.remote.execute(cmd));

      case 4:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 5:
        context$1$0.next = 7;
        return _regeneratorRuntime.awrap(this.getNativePageSource());

      case 7:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 8:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

helpers.getNativePageSource = function callee$0$0() {
  var method, endpoint, srcTree, src;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        method = 'POST';
        endpoint = '/source';
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap(this.proxyCommand(endpoint, method));

      case 4:
        srcTree = context$1$0.sent;
        src = getSourceXml(getTreeForXML(srcTree));
        return context$1$0.abrupt('return', src);

      case 7:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

// { isEnabled: '1',
//   isVisible: '1',
//   frame: '{{0, 0}, {375, 667}}',
//   children:
//    [ { isEnabled: '1',
//        isVisible: '1',
//        frame: '{{0, 0}, {375, 667}}',
//        children: [],
//        rect: { x: 0, y: 0, width: 375, height: 667 },
//        value: null,
//        label: null,
//        type: 'Other',
//        name: null,
//        rawIdentifier: null },
//   rect: { origin: { x: 0, y: 0 }, size: { width: 375, height: 667 } },
//   value: null,
//   label: 'UICatalog',
//   type: 'Application',
//   name: 'UICatalog',
//   rawIdentifier: null }
function getTreeForXML(srcTree) {
  function getTree(element, elementIndex, parentPath) {
    var curPath = parentPath + '/' + elementIndex;
    var rect = element.rect || {};
    var subtree = {
      '@': {
        name: element.name || '',
        label: element.label || '',
        value: element.value,
        dom: null,
        enabled: parseInt(element.isEnabled, 10) === 1,
        valid: true,
        visible: parseInt(element.isVisible, 10) === 1,
        hint: element.hint || '',
        path: curPath,
        x: rect.x,
        y: rect.y,
        width: rect.width,
        height: rect.height
      },
      '>': []
    };
    for (var i = 0; i < (element.children || []).length; i++) {
      subtree['>'].push(getTree(element.children[i], i, curPath));
    }
    return _defineProperty({}, 'XCUIElementType' + element.type, subtree);
  }
  var tree = getTree(srcTree.tree, 0, '');
  return tree;
}

function getSourceXml(jsonSource) {
  return (0, _js2xmlparser22['default'])("AppiumAUT", jsonSource, {
    wrapArray: { enabled: false, elementName: 'element' },
    declaration: { include: true },
    prettyPrinting: { indentString: '    ' }
  });
}

_Object$assign(extensions, commands, helpers);
exports.commands = commands;
exports.helpers = helpers;
exports['default'] = extensions;

// translate the returned value into what Appium expects
