// appium-windows-driver/build/lib/driver.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _slicedToArray = require('babel-runtime/helpers/sliced-to-array')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _appiumBaseDriver = require('appium-base-driver');

var _appiumSupport = require('appium-support');

var _winappdriver = require('./winappdriver');

var _winappdriver2 = _interopRequireDefault(_winappdriver);

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var WINAPPDRIVER_PORT = 4723;
var WINAPPDRIVER_HOST = "localhost";

// Appium instantiates this class

var WindowsDriver = (function (_BaseDriver) {
  _inherits(WindowsDriver, _BaseDriver);

  function WindowsDriver() {
    var opts = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var shouldValidateCaps = arguments.length <= 1 || arguments[1] === undefined ? true : arguments[1];

    _classCallCheck(this, WindowsDriver);

    _get(Object.getPrototypeOf(WindowsDriver.prototype), 'constructor', this).call(this, opts, shouldValidateCaps);
    this.jwpProxyActive = false;
    this.opts.port = opts.port || WINAPPDRIVER_PORT;
    this.opts.address = opts.address || WINAPPDRIVER_HOST;
  }

  _createClass(WindowsDriver, [{
    key: 'createSession',
    value: function createSession(caps) {
      var sessionId, _ref, _ref2;

      return _regeneratorRuntime.async(function createSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (_appiumSupport.system.isWindows()) {
              context$2$0.next = 2;
              break;
            }

            throw new Error("WinAppDriver tests only run on Windows");

          case 2:
            context$2$0.prev = 2;
            sessionId = undefined;
            context$2$0.next = 6;
            return _regeneratorRuntime.awrap(_get(Object.getPrototypeOf(WindowsDriver.prototype), 'createSession', this).call(this, caps));

          case 6:
            _ref = context$2$0.sent;
            _ref2 = _slicedToArray(_ref, 1);
            sessionId = _ref2[0];
            context$2$0.next = 11;
            return _regeneratorRuntime.awrap(this.startWinAppDriverSession());

          case 11:
            return context$2$0.abrupt('return', [sessionId, caps]);

          case 14:
            context$2$0.prev = 14;
            context$2$0.t0 = context$2$0['catch'](2);
            context$2$0.next = 18;
            return _regeneratorRuntime.awrap(this.deleteSession());

          case 18:
            throw context$2$0.t0;

          case 19:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[2, 14]]);
    }
  }, {
    key: 'startWinAppDriverSession',
    value: function startWinAppDriverSession() {
      return _regeneratorRuntime.async(function startWinAppDriverSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            this.winAppDriver = new _winappdriver2['default']({
              app: this.opts.app
            });
            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(this.winAppDriver.start());

          case 3:
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap(this.winAppDriver.startSession(this.caps));

          case 5:
            this.proxyReqRes = this.winAppDriver.proxyReqRes.bind(this.winAppDriver);
            // now that everything has started successfully, turn on proxying so all
            // subsequent session requests go straight to/from WinAppDriver
            this.jwpProxyActive = true;

          case 7:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'deleteSession',
    value: function deleteSession() {
      return _regeneratorRuntime.async(function deleteSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            _logger2['default'].debug('Deleting WinAppDriver session');

            if (!(this.winAppDriver && this.jwpProxyActive)) {
              context$2$0.next = 7;
              break;
            }

            context$2$0.next = 4;
            return _regeneratorRuntime.awrap(this.winAppDriver.deleteSession());

          case 4:
            context$2$0.next = 6;
            return _regeneratorRuntime.awrap(this.winAppDriver.stop());

          case 6:
            this.winAppDriver = null;

          case 7:
            this.jwpProxyActive = false;
            context$2$0.next = 10;
            return _regeneratorRuntime.awrap(_get(Object.getPrototypeOf(WindowsDriver.prototype), 'deleteSession', this).call(this));

          case 10:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'proxyActive',
    value: function proxyActive() {
      // we always have an active proxy to the WinAppDriver server
      return true;
    }
  }, {
    key: 'canProxy',
    value: function canProxy() {
      // we can always proxy to the WinAppDriver server
      return true;
    }
  }]);

  return WindowsDriver;
})(_appiumBaseDriver.BaseDriver);

exports.WindowsDriver = WindowsDriver;
exports['default'] = WindowsDriver;
